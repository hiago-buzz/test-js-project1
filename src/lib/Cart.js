import { find as findLodash, remove as removeLodash } from 'lodash';

export default class Cart {
  items = [];

  add(item) {
    const itemToFind = { product: item.product };

    if (findLodash(this.items, itemToFind)) {
      removeLodash(this.items, itemToFind);
    }

    this.items.push(item);
  }

  remove(product) {
    removeLodash(this.items, { product: product });
  }

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc + item.quantity * item.product.price;
    }, 0);
  }

  checkout() {
    return {
      total: this.getTotal(),
      items: this.items,
    };
  }
}
