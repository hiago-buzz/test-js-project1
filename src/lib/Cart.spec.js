import Cart from './Cart';

describe('Cart', () => {
  let cart;

  let product = {
    title: 'Adidas tennis',
    price: 2000,
  };

  let product2 = {
    title: 'Nike tennis',
    price: 4000,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe('getTotal()', () => {
    it('should return 0 when getTotal() is executed in a newly created instance', () => {
      expect(cart.getTotal()).toEqual(0);
    });

    it('should multiply quantity and price and receive the total amount', () => {
      cart.add({
        product,
        quantity: 2,
      });

      expect(cart.getTotal()).toEqual(4000);
    });

    it('should ensure no more than on product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });

      expect(cart.getTotal()).toEqual(2000);
    });

    it('should update total when a product gets included and removed', () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal()).toEqual(4000);
    });
  });

  describe('checkout', () => {
    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product: product2,
        quantity: 5,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });
  });
});
