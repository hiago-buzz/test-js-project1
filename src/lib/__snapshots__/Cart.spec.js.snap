// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout should return an object with the total and the list of items 1`] = `
{
  "items": [
    {
      "product": {
        "price": 2000,
        "title": "Adidas tennis",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 4000,
        "title": "Nike tennis",
      },
      "quantity": 5,
    },
  ],
  "total": 24000,
}
`;
